rfd/0098-securing-docker-images.md

```diff
@@ -0,0 +1,181 @@
+---
+authors: Trent Clarke (trent@goteleport.com)
+state: draft
+---
+
+# RFD 0098 - Delivering secure Teleport OCI Images
+
+# Required Approvers
+* Engineering: @r0mant
+* Security: @reed
+* Product: (@xinding33 || @klizhentas)
+
+## What
+
+This RFD discusses structures and processes to increase the security of the OCI 
+images we use to deliver Teleport to our customers.
+
+## Why
+
+One of our shipping channels is a Docker image. As the provider of that docker
+image we are (at least partialy) responsible for everything in it, not just
+Teleport. We should not ship vulnerabilities to our clients, even if those 
+vulnerabilities are not directly in our software.
+
+## Details
+
+### Goals
+
+For the sake of this RFD, I will define _delivering a secure image_ as 
+
+> Reliably producing an OCI image that:
+>
+>
+>  1. a priori is unlikely to contain vulnerabilities in and of itself, 
+>  2. contains software of known provenance,
+>  3. has no warnings or vulnerabilities flagged when run through a reputable 
+>     scanner at the time of creation, and
+>  4. flags vulnerabilities even after creation, either in Teleport itself or 
+>     a dependency.
+
+### Non-goals
+
+This RFD will not discuss building images for and/or running Teleport on
+platforms other than Linux.
+
+### Approach
+
+This RFD describes a 2-pronged approach for meeting the above goals:
+
+   1. Switching to distroless base images, and 
+   2. Automated, ongoing monitoring of images using a scanning service
+
+### 1. Distroless Images
+
+Distroless images contain only an application and the minimal set of
+dependencies for it. Google offers several base images that contain minimal 
+Linux distribution that we can use as a starting point.
+
+Switching to distroless images drastically reduces the number of software components 
+we ship as part of a Teleport distribution. This both reduces the size of the potential
+attack surface, and reduces the potential for high-noise reports from automated
+scanners. 
+
+The Google-supplied distroless base images also [provide a mechanism for verifying the
+provenance of a given image](https://github.com/GoogleContainerTools/distroless#how-do-i-verify-distroless-images) 
+using `cosign` and a public key. Stronger, SLSA-2 level guarantees [can be verified with additional
+tooling](https://security.googleblog.com/2021/09/distroless-builds-are-now-slsa-2.html). 
+
+**Note** that we are _already_  using Distroless images to distribute some Teleport 
+plugins. This would extend their use to Teleport proper.
+
+### 2. Ongoing scanning
+
+Using an _ongoing_ automated scanner means that we do not just check for 
+vulnerabilities at image creation time, instead we _continually_ scan for any vulnerabilities 
+that may be discovered during the image's lifetime.
```

Comment:
What do we define the lifetime as?

Reply:
For the sake of argument I am defining it as:

> _lifetime_ is the period of time from publishing the image until until either:
> * the image is replaced with a newer verion of the image, or
> * the support lifetime for the verion of Teleport on that  image expires (i.e. falls
  out of our 3-version support window)